src/Transients.cc

```diff
@@ -336,25 +329,30 @@ Transients::readers(const StoreEntry &e) const
 void
 Transients::markForUnlink(StoreEntry &e)
 {
-    unlink(e);
+    assert(e.key);
+    if (e.hasTransients())
+        abandon(e);
+    else
+        unlinkByKeyIfFound(reinterpret_cast<const cache_key*>(e.key));
 }
 
 void
-Transients::unlink(StoreEntry &e)
+Transients::unlinkByKeyIfFound(const cache_key *key)
```

Comment:
I replaced with Store::CacheKey, only if the corresponding context required information, provided by it.  For now, am afraid to get involved into unrelated changes, because there are many other methods (indirectly connected with that you mentioned) which should be fixed too.